### PR TITLE
Fix performance issue in `immutable.HashMap#concat`

### LINKED ISSUE #13039: `immutable.HashMap.concat` concatenates twice for no apparent reason
Scala version: 2.13.14, prob'ly also in Scala 3
Not a bug in the sense that it doesn't produce wrong results, only a minor but serious ugliness
```scala
case hm: HashMap[K, V1] =>
      if (isEmpty) hm
      else {
        val newNode = rootNode.concat(hm.rootNode, 0)
        if (newNode eq hm.rootNode) hm
        else newHashMapOrThis(rootNode.concat(hm.rootNode, 0))
      }
```
For comparison, `HashSet` using the same implementation:
```Scala
      case hs: HashSet[A] =>
        if (isEmpty) hs
        else {
          val newNode = rootNode.concat(hs.rootNode, 0)
          if (newNode eq hs.rootNode) hs
          else newHashSetOrThis(newNode)
        }
```
## Problem

That operation ain't that cheap to repeat it, especially that it was amusingly done in order to gain a minor performance benefit.
Yes, 'questions aren't bug reports', but if you know what's the complexity of `concat` I'd love to hear it, it's a bit hard to follow.

### PR DESCRIPTION
**Pull request by som-snytt**
The suggested fix.

Fixes scala/bug#13039

### CODE REVIEW

**Review by OndrejSpanel — APPROVED** [review]

**Comment by som-snytt** [comment]
I just did this. Moreover, as I typed in the edit, it crossed my distracted mind that I'd just seen this edit and how I'd never make the same mistake, which I did do about two seconds later. I noticed it because my test didn't turn green, as I did when I realized my folly.

**Comment by SethTisue** [comment]
review by @scala/collections ?
